server/service/microsoft_mdm.go

```diff
@@ -1188,31 +1194,104 @@ func (svc *Service) GetMDMWindowsTOSContent(ctx context.Context, redirectUri str
 }
 
 // isTrustedRequest checks if the incoming request was sent from MDM enrolled device
-func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncML) error {
+func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncML, reqCerts []*x509.Certificate) error {
 	if reqSyncML == nil {
 		return fleet.NewInvalidArgumentError("syncml req message", "message is not present")
 	}
 
-	// TODO - Security checks that need to be implemented
-	// - TLS based auth
-	// - Device auth based on Source/LocURI DeviceID information (this should be present on Enrollment DB)
+	// Checking if calling request is coming from an already MDM enrolled device
+	deviceID, err := reqSyncML.GetSource()
+	if err != nil || deviceID == "" {
+		return fmt.Errorf("invalid SyncML message %w", err)
+	}
+
+	enrolledDevice, err := svc.ds.MDMWindowsGetEnrolledDeviceWithDeviceID(ctx, deviceID)
+	if err != nil || enrolledDevice == nil {
+		return errors.New("device was not MDM enrolled")
+	}
 
-	return nil
+	// Check if TLS certs contains device ID on its common name
+	if len(reqCerts) > 0 {
+		for _, reqCert := range reqCerts {
+			if strings.Contains(reqCert.Subject.CommonName, deviceID) {
+				return nil
+			}
+		}
+	}
+
+	// TODO: Latest version of the MDM client stack don't populate TLS.PeerCertificates array
+	// This is a temporary workaround to allow the management request to proceed
+	// Transport-level security should be replaced for Application-level security
+	// Transport-level security is defined in the MS-MDM spec in section 1.3.1
+	// On the other hand, Application-level security is defined here
+	// https://www.openmobilealliance.org/release/DM/V1_2_1-20080617-A/OMA-TS-DM_Security-V1_2_1-20080617-A.pdf
+	// The initial values for Application-level security configuration are defined in the
+	// WAP Profile blob that is sent to the device during the enrollment process. Example below
+	//	<characteristic type="APPAUTH">
+	//		<parm name="AAUTHLEVEL" value="CLIENT"/>
+	//		<parm name="AAUTHTYPE" value="DIGEST"/>
+	//		<parm name="AAUTHSECRET" value="2jsidqgffx"/>
+	//		<parm name="AAUTHDATA" value="aGVsbG8gd29ybGQ="/>
+	//	</characteristic>
+	//	<characteristic type="APPAUTH">
+	//		<parm name="AAUTHLEVEL" value="APPSRV"/>
+	//		<parm name="AAUTHTYPE" value="DIGEST"/>
+	//		<parm name="AAUTHNAME" value="43f8bf591b8557346021"/>
+	//		<parm name="AAUTHSECRET" value="crbr3w2cab"/>
+	//		<parm name="AAUTHDATA" value="aGVsbG8gd29ybGQ="/>
+	//	</characteristic>
+
+	if len(reqCerts) == 0 {
```

Comment:
So in this case, we're relying only on shared knowledge of the device id as the basis for trust? And any person that knows the device id of an enrolled device would be able to to make MDM client requests and receive the command payloads from Fleet? Is it possible for the command payloads to contain network credentials or other secrets?

Reply:
@gillespi314 yes, if the MDM client stack does not populate the `TLS.PeerCertificates` information on the HTTP request with certificate information provisioned during enrollment, we are only relying on the DeviceID to authenticate the device and hence suffer from the possibility of a reply attack as you mentioned.

There is a limitation with using the transport-level security approach defined in Section 1.3.1 of the spec because the OS is not sending this information. I've created this subtask to look into the legacy authentication mechanism mentioned on the MS-MDM spec. The subtask [here](https://github.com/fleetdm/fleet/issues/14662) has more details on this.

Reply:
@gillespi314 this is not going to be the final implementation, let's move forward with this while we're still under a feature flag and give the authentication the attention it deserves in a separate ticket.

@marcosd4h could you please add specifics about the windows versions with problems in the linked issue? just so we know what to test.

Reply:
Sure! This was found with Windows 10 RS5 (Version 1809 - Major Build 17763). I'm using this version mostly because this was the first version where MDM enrollment and management clients were introduced according to [this](https://learn.microsoft.com/en-us/windows/client-management/new-in-windows-mdm-enrollment-management)

I spent more time this morning looking at this, and I found a doc from Microsoft with information that was not in the spec. See [here](https://learn.microsoft.com/en-us/windows/client-management/server-requirements-windows-mdm) 

I need to try this out, but it seems that for the TLS cert to go through, the certificate used to provision the device identity cannot be a self-signed cert.

![image](https://github.com/fleetdm/fleet/assets/17311749/198ee63c-5579-435c-8ecb-0cc3b2c2c325)